Steiner Tree Feature

Nachiket18 opened the pull request:
The functionality that is being developed is calculating the Steiner tree in a graph. Steiner Tree is created on basis of subset of terminals in the graphs with the condition that length of the tree should be minimum.

We are following the approach suggested in this paper - https://arxiv.org/abs/1710.00668

Also, we are taking reference of an FPT algorithm known as Dreyfus- Wagner algorithm - https://onlinelibrary.wiley.com/doi/10.1002/net.3230010302

----

Edit by @szhorvat:

TODO list to be completed only _after_ the first review:

 - [x] interruptability
 - [ ] functions.yaml
 - [ ] documentation
 - [x] checks on input parameters
 - [ ] overflow checks
 - [x] decide whether `weights=NULL` should be accepted (meaning identical weights) and how this is best handled

szhorvat commented:
This is a good starting point. I close the other PR (the comments won't disappear).

The next step now is to fix all the errors that prevent the code from compiling. You will not be able to check if it works correctly until it compiles.

Are you comfortable with [compiling igraph](https://igraph.org/c/html/develop/igraph-Installation.html) on your own computer _without_ these changes? Do you know how to run the tests?

Nachiket18 commented:
> This is a good starting point. I closed the other PR (the comments won't disappear).
>  
> The next step now is to fix all the errors that prevent the code from compiling. You will not be able to check if it works correctly until it compiles.
> 
> Are you comfortable with [compiling igraph](https://igraph.org/c/html/develop/igraph-Installation.html) on your own computer _without_ these changes? Do you know how to run the tests? You should be using the very same steps to compile and test the function implemented in this PR.

Thanks. I was able to compile the code using the instructions which specifies the usage for cmake and later the command cmake --build <dir>. Earlier there were compile time errors but at present the code ( including the Steiner tree change) is getting build without any errors. 

I am however, confused about running test case. I did create create a test case and updated the CmakeLists.txt but running it and understanding why the CI is failing is proving little difficult for me.

Nachiket18 commented:
I was able to find out the error in the test file and one with unused variable in steiner.cpp which I corrected and pushed it on the repository. The CI tests didn't run for the last two commit. Could you help me with this?

szhorvat commented on include/igraph_paths.h:
```diff
@@ -47,9 +47,15 @@ IGRAPH_EXPORT igraph_error_t igraph_diameter_dijkstra(const igraph_t *graph,
                                         const igraph_vs_t from, const igraph_vs_t to,
                                         igraph_neimode_t mode);
+										
+IGRAPH_EXPORT int igraph_steiner_dreyfus_wagner(const igraph_t *graph,
```
Error codes must now be `igraph_error_t`, not `int`.

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,72 @@
+    
+    igraph_vector_t steiner_terminals;
+    igraph_vector_t weights_empty, weights_lm;
+    igraph_integer_t dreyfus_wagner_out;
```
This variable is used for an error code, so it should be `igraph_error_t`. (But do you need to store an error code at all?)

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,225 @@
+ return key;
+}
+
+igraph_integer_t igraph_steiner_dreyfus_wagner(const igraph_t graph,
```
This does not match the prototype in the header (this is the cause of the linking failure you see on CI).

szhorvat commented:
Can you please format the source files for readability? Use consistent indentation. Your editor should have an auto-format command.

Eventually, the formatting will need to match that of the rest of igraph, but for now, let's just have something that is consistent and readable, and we can worry matching the rest of the codebase late.

Nachiket18 commented:
Thanks @szhorvat 

I made the changes that you had suggested. Could you review the issue that is caused at IGRAPH_ASSERT macro?

For next steps, I have something that I would need to discuss with you -

The igraph_steiner_dreyfus_wagner function calculates the minimum length of the Steiner tree for smaller graph as it's FPT algorithm for NP-complete problem. Should the code be designed in such a way that it outputs tree as result or would output have tree and minimum length?

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,69 @@
+    IGRAPH_ASSERT(igraph_steiner_dreyfus_wagner(&g_empty, steiner_terminals, IGRAPH_ALL, &weights_empty) == IGRAPH_SUCCESS);
+
+    printf("Un-Directed graph with loops and multi-edges, select none:\n");
+    IGRAPH_ASSERT(igraph_steiner_dreyfus_wagner(&g_lm, steiner_terminals, IGRAPH_ALL, &weights_lm == IGRAPH_SUCCESS);
```
Compilation fails because of a missing closing parenthesis:

```suggestion
    IGRAPH_ASSERT(igraph_steiner_dreyfus_wagner(&g_lm, steiner_terminals, IGRAPH_ALL, &weights_lm) == IGRAPH_SUCCESS);
```

szhorvat commented:
> Should the code be designed in such a way that it outputs tree as result or would output have tree and minimum length?

It would be nice if the function could (optionally) provide both.

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,69 @@
+    
+    igraph_vector_init(&weights_empty, 0);
+
+    igraph_vector_init(&steiner_terminals, 0);
```
This vector is initialized to size zero, but later you are trying to write into it. There are several instances of this.

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,69 @@
+
+    igraph_vector_init(&steiner_terminals, 0);
+
+    VECTOR(steiner_terminals)
+    [1] = 2;
+    VECTOR(steiner_terminals)
+    [2] = 4;
```
Please do not put the `[index]` on a different line.

I'm not sure about intention here, but note that indexing in C is zero-based, not 1-based. You are starting at 1.

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,69 @@
+    VECTOR(steiner_terminals)
+    [2] = 4;
+
+    igraph_vector_init_int(&weights_lm, 0);
```
Did you mean to use `vector_init` instead of `vector_init_int`? Also, this vector is initialized to size zero as well, then you write into it.

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,69 @@
+
+    igraph_small(&g_empty, 0, 0, -1);
+
+    igraph_small(&g_lm, 5, IGRAPH_UNDIRECTED, 0, 1, 0, 2, 1, 2, 1, 3, 1, 4, 2, 3, 3, 4 - 1);
```
This missing comma would lead to a crash.

```suggestion
    igraph_small(&g_lm, 5, IGRAPH_UNDIRECTED, 0, 1, 0, 2, 1, 2, 1, 3, 1, 4, 2, 3, 3, 4, -1);
```

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,69 @@
+
+int main()
+{
+    igraph_t g_empty, g_empty_dir, g_lm;
```
Unused variable will cause a warning on some platforms, which leads to CI failure.

```suggestion
    igraph_t g_empty, g_lm;
```

szhorvat reviewed (commented):
When there is a crash, check the output from `linux_static_vendored`. This gives you the most information.

https://dev.azure.com/igraph-team/igraph/_build/results?buildId=2416&view=logs&j=1a52b443-5fb8-53f1-821f-633f5d086e98&t=16950513-9d69-54bb-da2c-179f24a13cca&l=370

I noted the cause of this crash, as well as some corrections that need to be made to avoid some other crashes.

Nachiket18 commented:
@szhorvat - Thanks for the suggestions. I made some silly mistakes in the test case file. sorry about that. For building the code I'm using the following commands -

$ cmake --build .
$ cmake --build . --target check

Which commands should I run to check the errors and resolve them locally? I will keep a note of the linux_static_vendored when the tests are run here on the platform.

szhorvat commented:
Yes, these commands are correct. If you use them you should see the test failing, just like on the CI right now. Does the test not fail for you locally?

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,61 @@
+    VECTOR(weights_lm)[5] = 2;
+    VECTOR(weights_lm)[6] = 7;
+
+    igraph_small(&g_empty, 0, 0, -1);
```
Here you can also use `igraph_empty()`.

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,61 @@
+
+    igraph_small(&g_empty, 0, 0, -1);
+
+    igraph_small(&g_lm, 5, IGRAPH_UNDIRECTED, 0, 1, 0, 2, 1, 2, 1, 3, 1, 4, 2, 3, 3, 4,1);
```
The `-1` terminator is missing, so this will cause a crash or other failure.

szhorvat replied:
```suggestion
    igraph_small(&g_lm, 5, IGRAPH_UNDIRECTED, 0, 1, 0, 2, 1, 2, 1, 3, 1, 4, 2, 3, 3, 4, -1);
```

szhorvat commented:
@Nachiket18 Let us know if you need more help with this. The immediate goal is to get the test suite to pass, and to make sure that you can run the tests locally on your computer.

I don't want to put too much pressure on you, so if you feel that you have no time right now, just let us know.

Nachiket18 commented:
@szhorvat - I made the changes that you had suggested. I was able to build the code using the instructions from the official documentation.  However, during the target check there was error generated. 

177/424 Test #177: test::igraph_steiner_tree_fpt......................***Failed    1.36 sec
Test exited abnormally with error: Segmentation fault
=========================================
CMake Error at /home/nachiket/igraph/etc/cmake/run_legacy_test.cmake:45 (message):
  Exiting test.

Normally, I would add print statements to debug the code but for this project what should be the way to resolve it?

ntamas commented:
Either add `printf` statements to the tests if that's convenient for you and run it manually from the command line (the test should be built in `test/test_igraph_steiner_tree_...` in the build folder -- just make sure you rebuild if after adding the `printf` statements), or hook it up to a debugger and run through it step by step. I don't know what IDE / OS you are using, I usually use `gdb` or `lldb` from the command line on macOS, but I'm oldschool :)

ntamas commented:
Resolved the conflicts with the upstream `develop` branch, FYI.

I haven't checked the PR thoroughly yet so this is _not_ a full review, but one thing that surely needs to be changed is that we have switched to using `igraph_vector_int_t` for variables that represents vertex indices so the function needs to be changed to use `igraph_vector_int_t` for vectors containing vertex indices. As for the distance matrix, that can probably stay as `igraph_matrix_t` because it can contain floating-point numbers in the weighted case, but infinite distances should be represented with `IGRAPH_INFINITY` and not `INT_MAX`.

szhorvat commented:
FYI I resolved conflict with `develop` again and made a minor fix to allow it to build again. You should pull the changes.

szhorvat commented:
If you haven't yet, take a look at this wiki page and the links within: https://github.com/igraph/igraph/wiki/Quickstart-for-new-contributors

As for debugging, building with AddressSanitizer (ASan) is useful. See here: https://github.com/igraph/igraph/wiki/Using-sanitizers-to-find-bugs

The `linux_static_vendored` CI run has ASan enabled and will give more informative error messages.  

https://dev.azure.com/igraph-team/igraph/_build/results?buildId=3578&view=logs&j=4c3a8381-b9de-5463-904f-918810769be8&t=e71559a0-54d3-55f3-64cc-74b67ad70398&l=400

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,211 @@
+		IGRAPH_ERROR("Weight vector length does not match", IGRAPH_EINVAL);
+	}
+
+	igraph_shortest_paths_johnson(graph, &distance, igraph_vss_all(), igraph_vss_all(), weights);
```
With all igraph functions, except constructors, any data structure that is passed in must be initialized. Here, `distance` not initialized, hence the segmentation fault.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,211 @@
+	return key;
+}
+
+igraph_error_t igraph_steiner_dreyfus_wagner(const igraph_t *graph,igraph_vector_t steiner_terminals,
```
`steiner_terminals` should be passed by reference, i.e. use a pointer. `igraph_vector_t *`. If you need to modify it, e.g. sort it, make a copy first.

szhorvat commented:
Let us know if you have more questions.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,211 @@
+	{
+		for (igraph_integer_t j = 0; j < igraph_vector_size(&steiner_vertices); j++)
+		{
+			igraph_matrix_set(&dp_cache, (igraph_integer_t)VECTOR(steiner_terminals)[i], (igraph_integer_t)VECTOR(steiner_vertices)[j], MATRIX(distance, (igraph_integer_t)VECTOR(steiner_terminals)[i], (igraph_integer_t)VECTOR(steiner_vertices)[j]));
```
You can assign element `i, j` of a matrix `m` as `MATRIX(m, i, j) = ...`

Nachiket18 commented:
> Let us know if you have more questions.

Thanks. I have done all the changes that you have specified and run the code locally. It's resulting into fatal error. What would be the correct way to locate the error as the error doesn't mention line number. Should I install debugger locally or should I just commit to repository and let the testing pipeline run so I can look into the issue here on GitHub ?

Nachiket18 commented:
@szhorvat - Thanks. I made the changes that were causing segmentation fault, changed the data type to 'igraph_vector_int_t' for steiner_terminals and also updated the test case. The test case is failing even though it passes through the function in the test case and fails in the end. We are looking at the results from linux_static_vendored pipeline and would try to fix it.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,251 @@
+		for (auto j = (*i).begin() ; j != (*i).end() ; j++)
+		{
+			if (j != (*i).begin()) { printf(", ");}
+			printf("%ld",*j);
```
This is causing a compilation error on macOS. The type of `*j`, which is `igraph_integer_t` is not necessarily the same as `long int`. Use `"%" IGRAPH_PRId` instead of `"%ld"`.

That said, all `printf`s will need to be removed (or hidden behind debug macros) before the PR can be merged. igraph is used from R, and CRAN policy is not to allow `printf` in the code.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,251 @@
+
+	if (igraph_vector_size(weights) != no_of_edges)
+	{
+		char error[1024] = {0};
+		sprintf(error,"Weight vector length does not match %ld vec size and %ld edges \n",igraph_vector_size(weights),no_of_edges);
+		IGRAPH_ERROR(error, IGRAPH_EINVAL);
```
Use `IGRAPH_ERRORF` instead of `sprintf` + `IGRAPH_ERROR`, and make sure to use `"%" IGRAPH_PRId` instead of `"%ld"` when printing an `igraph_integer_t`.

RonakGSahu replied:
Hi, I tried using `"%" IGRAPH_PRId` but it doesn't recognize it. Will you be able to show an example of how it's used? I couldn't find it in the develop section Reference Manual.

ntamas replied:
`IGRAPH_PRId` is defined here: https://github.com/igraph/igraph/blob/develop/include/igraph_types.h#L64

There are plenty of usage examples for `IGRAPH_PRId` in the codebase, for instance, here:

https://github.com/igraph/igraph/blob/develop/tests/unit/igraph_rng_get_integer.c#L33

You made a good point with the documentation, though - I think it's not mentioned anywhere in the docs yet, only in the examples, so we should fix that.

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,118 @@
+                                                -1);
+    
+    //printf("No vertices, not directed:\n");
+    IGRAPH_ASSERT(igraph_steiner_dreyfus_wagner(&g_empty,&steiner_terminals_null, IGRAPH_UNDIRECTED, &weights_empty) == IGRAPH_SUCCESS);
```
Tip: you don't need to assert a return of `IGRAPH_SUCCESS` in tests _unless_ you changed the default error handler. If there is an error, the program will abort by default, and the test will fail.

Nachiket18 replied:
Okay. I had looked into other test cases and since they were using it, I thought it is necessary. But I will make sure to change it

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,251 @@
+#include "igraph_error.h"
+#include "igraph.h"
+
+#include <string.h>
```
In C++, `srtring.h` should be `cstring`, `math.h` should be `cmath`, etc.

Nachiket18 commented:
> > infinite distances should be represented with `IGRAPH_INFINITY` and not `INT_MAX`.
> 
> Just highlighting this comment from @ntamas, as this needs to be addressed.
> 
> There are some parts of the algorithm that are not covered by the tests. Can you make sure that everything is covered, except some checks on input parameters or other error checks?
> 
> Let us know when this is ready for the first round of review.

Yes. We would write test case for checking the logic of the algorithm by printing and comparing the value for minimum steiner tree distance with one that's expected.

szhorvat commented:
One more comment is that it will be useful to merge `develop` into this branch again, to prevent the two branches from diverging too much. This should also fix the coverage report, which incorrectly claims that this PR reduces coverage in many files.

szhorvat commented:
It's great to see this PR moving forward.

To reiterate, the immediate goals are:

 - [ ] Merge the `develop` branch into this branch.
 - [ ] Fix test failures on all platforms. Let us know if it's unclear how to deal with a failure.
 - [ ] Ensure full test coverage of all parts of the code that implement the algorithm.  You can see coverage [here](https://app.codecov.io/gh/igraph/igraph/compare/1939/diff?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr%20comments&utm_term=igraph)  It is not necessary to cover error checks or debug print functions, just the core algorithm.

Before the review, it would be great to reformat the code to fit with the rest of the igraph codebase. We use an [editorconfig](https://editorconfig.org/) file to help with this. If you're not sure how to do this efficiently, let me know and I will run it through an auto-formatter.

Once these are done, we can start an earnest review.

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,119 @@
+    igraph_vector_int_init(&steiner_terminals, 4);
+    igraph_vector_int_init(&steiner_terminals_null, 0);
+
+    VECTOR(steiner_terminals)[0] = 0;
+    VECTOR(steiner_terminals)[1] = 1;
+    VECTOR(steiner_terminals)[2] = 2;
+    VECTOR(steiner_terminals)[3] = 3;
```
Tip: initializing a vector to a given value is more easily done with `igraph_vector_int_init_int()` or `igraph_vector_int_init_real()`. For example, `igraph_vector_int_init_int(&steiner_terminals, 4 /* element count */, 0, 1, 2, 3);` You can grep the contets of the `tests` directory for example uses of these functions.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,252 @@
+			}
+		}
+	}
+	igraph_integer_t u = IGRAPH_INTEGER_MAX;
+	igraph_integer_t v = IGRAPH_INTEGER_MAX;
```
Are these variables representing a distance? If so, can you please choose a different name for them? `u` and `v` are traditionally used to denote vertices (in igraph: vertiex IDs).

If these are distances, should they not be of type `igraph_real_t`? In the weighted case distances are not necessarily integer valued.

szhorvat replied:
Please click "Resolve conversation" under all review comments that have been addressed (unless there's anything left to do). This way we can keep track of what still needs work.

RonakGSahu commented:
@szhorvat I tried to fix as many issues as possible and am at my last 2 as per what I see the last few ones deal with 

1. macOS -> Johnsons being deprecated which causes Mac to fail at this moment.
2. Windows(both) - > it says the `min` function doesn't belong to the namespace `std` but when I click the definition it says it is and I don't know how to change it. And this is the only error left in windows at this moment.

Do you know how to deal with this cross-compatibility before I move on with the functionality and its test case? Thanks

szhorvat commented:
When a function is deprecated, look it up in the sources and see what it was replaced with. In this case it is `igraph_distances_johnson()`.

As for `min()`, let us look up which header declares it: https://en.cppreference.com/w/cpp/algorithm/min  It is `<algorithm>`.  With some of the standard C++ library implementations, some other headers _may_ declare `min` as well, which is why you see no failure on Windows. But only `<algorithm>` is _guaranteed_ to declare it, thus the only standards-compliant way to proceed is to include it.

Nachiket18 commented:
We were able to build and test the functionality. Our test case includes a graph given in the research paper and output that code is generating is same as what it's mentioned in the paper. In order to ensure correctness, how many such test cases need to be implemented ?
Meanwhile, I'll start working on documentation.

szhorvat commented:
There's a wiki page that has hints on what to pay attention to when testing, as well as a few other things that we'll need to get done eventually.

https://github.com/igraph/igraph/wiki/Checklist-for-new-%28and-old%29-functions

 - Consider the case of 0 and 1 vertex graphs. What is a reasonable result and does the function return that result?
 - What if the input is disconnected?
 - Does your implementation handle self-loops and multi-edges? Self-loops are easy as they can be ignored, but multi-edges are not. If the current implementation does not handle them, can be changed to support them?
 - etc.

szhorvat commented:
Don't forget to merge the `develop` branch into this one.

szhorvat commented:
It looks like the function currently only returns the total weights of the Steiner tree. Can you return the tree itself as well as a vector of edge indices, similar to how `igraph_minimum_spanning_tree()` does it?

Nachiket18 commented:
> It looks like the function currently only returns the total weights of the Steiner tree. Can you return the tree itself as well as a vector of edge indices, similar to how `igraph_minimum_spanning_tree()` does it?

Yes. We are figuring out how we are going to do.

Nachiket18 commented:
> > It looks like the function currently only returns the total weights of the Steiner tree. Can you return the tree itself as well as a vector of edge indices, similar to how `igraph_minimum_spanning_tree()` does it?
> 
> Yes. We are figuring out how we are going to do.

@szhorvat  - We are implementing procedure for generating the Steiner Tree. This involves getting path from a particular vertex to another and this process repeats several times. Now our doubt is that how do we get a shortest path from a particular vertex to another in the graph. We have all pair shortest path calculated by Johnson's algorithm which gives us a matrix as a result but not actual paths.

RonakGSahu commented:
> > > It looks like the function currently only returns the total weights of the Steiner tree. Can you return the tree itself as well as a vector of edge indices, similar to how `igraph_minimum_spanning_tree()` does it?
> > 
> > 
> > Yes. We are figuring out how we are going to do.
> 
> @szhorvat - We are implementing procedure for generating the Steiner Tree. This involves getting path from a particular vertex to another and this process repeats several times. Now our doubt is that how do we get a shortest path from a particular vertex to another in the graph. We have all pair shortest path calculated by Johnson's algorithm which gives us a matrix as a result but not actual paths.

Adding to this, we tried looking through the reference manual but we can't seem to find a function that returns the graph (or the set of edges) just the weighted length of it.

szhorvat commented:
The functions which return shortest paths (instead of distances) are all called `..._get_shorest_paths()` and they are documented here:

https://igraph.org/c/html/develop/igraph-Structural.html#shortest-path-related-functions

There isn't one that corresponds to Johnson's algorithm at the moment.

Note that `get_shortest_path()` (singular) is just a wrapper for getting the path between two specific vertices, but it does not offer efficiency gains.

----

Do you mean that you repeatedly need to query for paths between `u` and `v`, even for the _same_ `(u, v)` pair in some cases?  If so, perhaps a good approach is to compute the shortest path tree (effectively returned through the `parents` and `inbound_edges` arguments of `get_shortest_paths()`) and when you do an `u -> v` query, retrieve the tree corresponding to `u`, and walk up on it from `v` back to `u`. This gets you the path quickly, in `O(pathlength)` time.

RonakGSahu commented:
Also, while going and trying to use the `igraph_get_shortest_path_djikstra` documentation from the manual we found that the documentation is wrong in one place. This is the Argument  `igraph_vector_int_t* edges` [link](https://igraph.org/c/html/develop/igraph-Structural.html#igraph_get_shortest_path_dijkstra). It says it needs an uninitialized pointer but this is wrong, it does need an initialized pointer to work with it ( Went and saw the function's code before we made this comment). Can somebody take a look as well and correct it?

ntamas commented:
Thanks for the heads up, the documentation of the `edges` argument for shortest path functions is now fixed in db22030c8

Nachiket18 commented:
> Also, while going and trying to use the `igraph_get_shortest_path_djikstra` documentation from the manual we found that the documentation is wrong in one place. This is the Argument `igraph_vector_int_t* edges` [link](https://igraph.org/c/html/develop/igraph-Structural.html#igraph_get_shortest_path_dijkstra). It says it needs an uninitialized pointer but this is wrong, it does need an initialized pointer to work with it ( Went and saw the function's code before we made this comment).



> The functions which return shortest paths (instead of distances) are all called `..._get_shorest_paths()` and they are documented here:
> 
> https://igraph.org/c/html/develop/igraph-Structural.html#shortest-path-related-functions
> 
> There isn't one that corresponds to Johnson's algorithm at the moment.
> 
> Note that `get_shortest_path()` (singular) is just a wrapper for getting the path between two specific vertices, but it does not offer efficiency gains.
> 
> Do you mean that you repeatedly need to query for paths between `u` and `v`, even for the _same_ `(u, v)` pair in some cases? If so, perhaps a good approach is to compute the shortest path tree (effectively returned through the `parents` and `inbound_edges` arguments of `get_shortest_paths()`) and when you do an `u -> v` query, retrieve the tree corresponding to `u`, and walk up on it from `v` back to `u`. This gets you the path quickly, in `O(pathlength)` time.

So we implemented the algorithm to generate tree along with shortest distance. What we did is that whenever we want to find path between to vertices u, v then we just passed those vertices to Igraph's Dijkstra's function, This is not the most efficient way to do it but right now at least the tree is generated.

szhorvat commented:
Don't forget to:

 - add comments to the code and explain what each function does where not entirely obvious
 - merge `develop` into this branch from time to time

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,135 @@
+
+/*
+   IGraph library.
+   Copyright (C) 2021  The igraph development team <igraph@igraph.org>
```
```suggestion
   Copyright (C) 2022  The igraph development team <igraph@igraph.org>
```

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,135 @@
+
+#include <igraph.h>
+#include "test_utilities.h"
+#include "igraph_memory.h"
+#include "igraph_types.h"
+#include "igraph_adjlist.h"
```
```suggestion
```

Tests should only include `<igraph.h>`

szhorvat replied:
You marked this as "resolved", but did not actually address the comment. The unnecessary headers were not removed.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,535 @@
+#include <climits>
+#include <vector>
+#include <set>
+#include <iostream>
```
`<iostream>`, as well as any terminal output by this function, should be removed before the PR can be merged.

If you use this for debugging, all debug code should be conditional on a macro being defined.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,535 @@
+	IGRAPH_FINALLY(igraph_matrix_destroy,&dp_cache);
+
+    igraph_matrix_fill(&dp_cache, IGRAPH_INFINITY);
+	for (long int i = 0 ; i  <  no_of_vertices; i++)
```
All indices (indexing into arrays, vertex or edge IDs) must be `igraph_integer_t`. Do not use `long int` or `int` without good reason.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,535 @@
+		for (long int j = 0 ; j  <  no_of_vertices; j++)
+		{
+			//std::cout << igraph_matrix_get(&distance, i,j) << " ";
+			igraph_matrix_set(&dp_cache,i,j,igraph_matrix_get(&distance,i,j));
```
Please use the `MATRIX` and `VECTOR` macros instead of `matrix_set/get` and `vector_set/get`.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,535 @@
+{
+	if (mode != IGRAPH_ALL)
+	{
+		std::cout << "Currently this function only supports undirected graphs while the graph's mode is not undirected." <<std::endl;
```
Use `IGRAPH_ERROR` or `IGRAPH_ERRORF` for errors. There must be no direct terminal output.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,535 @@
+	if (mode != IGRAPH_ALL)
+	{
+		std::cout << "Currently this function only supports undirected graphs while the graph's mode is not undirected." <<std::endl;
+		return IGRAPH_FAILURE;
```
Please avoid returning error code directly whenever possible, with the exception of `IGRAPH_SUCCESS`. Instead, use the `IGRAPH_ERROR` and `IGRAPH_ERRORF` macros which call error handler in addition to returning a code.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,535 @@
+	}
+	
+	
+	igraph_integer_t no_of_vertices = (igraph_integer_t)igraph_vcount(graph);
```
```suggestion
	igraph_integer_t no_of_vertices = igraph_vcount(graph);
```

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,535 @@
+	if (no_of_vertices == 0 || (no_of_vertices == 1)) //graph is empty
+	{
+		*res = 0;
+		return IGRAPH_FAILURE;
```
Do not return a non-success error code without also calling the error handler. In other words, use the `IGRAPH_ERROR` or `IGRAPH_ERRORF` macros.

szhorvat reviewed (commented):
I pointed out a few small things I noticed after a superficial skim. Hopefully this will help us get this PR to a reviewable state faster. This is not a real review yet. We'll wait until you say that the PR is ready for that.

szhorvat commented on tests/unit/igraph_steiner_tree_fpt.c:
```diff
@@ -0,0 +1,135 @@
+    printf("%.2f\n",val2);
+    IGRAPH_ASSERT(val2 == 5);
+    
+    for (igraph_integer_t i=0; i < igraph_vector_int_size(&res_tree_1); ++i){
+      printf("%"IGRAPH_PRId",",VECTOR(res_tree_1)[i]);
+    }
```
Please use `print_vector_int` from `test_utilities.h` to print the contents of an `igraph_vector_int_t`.

RonakGSahu commented:
Is there any way to test `igraph_error_t` outputs besides `IGRAPH_SUCCESS` I tried to test normally but it calls the abort function prints the correct statement but it then quits the test case file? I looked at other test cases and tried to imitate it but it doesn't seem to be working.

ntamas commented:
`igraph_set_error_handler(igraph_error_handler_ignore)` will prevent the error handler from aborting the app so you can do whatever you want with the return code.

szhorvat commented:
There is a convenient `CHECK_ERROR` macro in `test_utilities.h` which will do what @ntamas described.

```
CHECK_ERROR(igraph_something(bad_argument), IGRAPH_EINVAL);
```

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,480 @@
+	{	
+		*res = 0;
+		IGRAPH_ERRORF("Weight vector length does not match %" IGRAPH_PRId "vec size and %" IGRAPH_PRId "edges \n",IGRAPH_FAILURE,igraph_vector_size(weights), no_of_edges);
+		return IGRAPH_FAILURE;
```
The `IGRAPH_ERRORF` macro already includes the `return` statement. It is not necessary to add an extra `return`.

Nachiket18 replied:
We have corrected it. Thanks

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,480 @@
+	if (igraph_vector_size(weights) != no_of_edges)
+	{	
+		*res = 0;
+		IGRAPH_ERRORF("Weight vector length does not match %" IGRAPH_PRId "vec size and %" IGRAPH_PRId "edges \n",IGRAPH_FAILURE,igraph_vector_size(weights), no_of_edges);
```
I noticed that currently you are only using the `IGRAPH_FAILURE` error code. When the input to the function is invalid, please use `IGRAPH_EINVAL` instead.

Additionally: There should be no newline character at the end of error messages and errors should be punctuated, as a full sentence. You will find the guidelines for error reporting [here](https://github.com/igraph/igraph/wiki/Error-reporting-guidelines). Fixing this is not a priority now, as I will go through the errors during the review and polish the phrasing anyway. I just wanted to let you know about this for the sake of future error messages you might add.

Nachiket18 replied:
Thanks. We made the change for IGRAPH_FAILURE

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,480 @@
+	}
+	if (igraph_vector_int_size(steiner_terminals) == no_of_vertices)
+	{
+		*res = 0;	
```
`res` is meant to contain the total weight of the Steiner tree, correct? If so, this should not be zero here.

Nachiket18 replied:
Yes. Thanks or pointing out, we have modified it.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,480 @@
+}
+
+
+igraph_error_t igraph_steiner_dreyfus_wagner(const igraph_t *graph,const igraph_vector_int_t* steiner_terminals,
+igraph_neimode_t mode, const igraph_vector_t *weights,igraph_real_t *res,igraph_vector_int_t *res_tree)
```
Currently you only implemented the undirected case. We will need to decide if this function will even be generalized to the directed case. If not, it will be better to remove the `mode` argument entirely and document that the function ignores edge directions.

Correct me if I'm wrong, but it seems to me that the directed Steiner tree problem is conceptually different from the undirected one. The variants I found start with a specific root node, which we don't need for the undirected case. Does the Dreyfus-Wagner algorithm generalize to the directed case at all? If not, we do not need to consider the future possibility of directed genertalizations and change the function signature accordingly.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,480 @@
+	igraph_integer_t no_of_edges = igraph_ecount(graph);
+
+	
+	if (no_of_vertices == 0 || (no_of_vertices == 1)) //graph is empty
```
Can we not generalize this function to handle the 0 and 1-vertex cases in a reasonable way? The spanning tree functions do. Instead of throwing an error, check that arguments are valid (`steiner_terminals` contains only valid vertex IDs, etc.) and return a reasonable result.

szhorvat reviewed (commented):
I added a few more small comments. Additionally, a few questions/comments:

 - I assume negative weights are not supported. The input should be checked for negative weights. See how `igraph_closeness()` does this for guidance.
 - Are zero weights supported?
 - What does the function do when presented with a disconnected graph and terminals that are in different connected components?

Nachiket18 commented:
> I added a few more small comments. Additionally, a few questions:
> 
> * I assume negative weights are not supported? If so, the input should be checked for negative weights. See how `igraph_closeness()` does this for guidance.
> * Are zero weights supported?
> * What does the function do when presented with a disconnected graph and terminals that are in different connected components?

1. Negative weights are not supported. So yes we would add that check.
2. The algorithm doesn't mention anything about zero weights but we are assuming they are non-zero
3.  I think the algorithm expects a connected graph and in case it's disconnected then we would add check to handle it.

szhorvat commented:
> The algorithm doesn't mention anything about zero weights but we are assuming they are non-zero

Can this restriction be relaxed? Which part of the code assumes that they are non-zero?



> I think the algorithm expects a connected graph and in case it's disconnected then we would add check to handle it.

Is it possible to relax this restriction as well and make the function work when all terminals are in the same component?

RonakGSahu commented:
I just realized that I don't have the same access as @Nachiket18 has for this PR as I found out I don't have the `Review Button` to press on. Is that normal I have Nachiket's igraph repository forked.

Nachiket18 commented:
> > The algorithm doesn't mention anything about zero weights but we are assuming they are non-zero
> 
> Can this restriction be relaxed? Which part of the code assumes that they are non-zero?
>  I will try to read the algorithm again to see if we can relax this restriction
> > I think the algorithm expects a connected graph and in case it's disconnected then we would add check to handle it.
> 
> Is it possible to relax this restriction as well and make the function work when all terminals are in the same component?
   It is possible to do that it would require some piece of code to identify if the Steiner terminals are in same connected component. Should we do this in this PR or should we make the modifications in later PR's?

szhorvat commented:
> It is possible to do that it would require some piece of code to identify if the Steiner terminals are in same connected component. Should we do this in this PR or should we make the modifications in later PR's?

You can use `igraph_connected_components()` to determine which component each vertex belongs to. It produces a `membership` vector. You can test if `u` and `v` are the same component using `membership[u] == membership[v]`.

If would be nice to do this in the same PR, as it's fairly easy, but it does not need to be before the first review.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,251 @@
+#include "igraph_types.h"
+#include "igraph_memory.h"
+#include "igraph_error.h"
+#include "igraph.h"
```
This still needs to be fixed.

szhorvat replied:
It seems that you misunderstood. What I meant was: do not use `#include "igraph.h"`. Instead, include only those sub-headers that are needed. Don't worry if there are a few too many, we'll remove the unneeded ones at the end. A rough approximation will suffice for now.

szhorvat commented:
Before starting a review, I would like the code formatted in the same style as the rest of igraph. You can do this most easily with the `astyle` utility. An `.astylerc` file is included in the repository. If you're not sure how to do this, let me know and I will push that change.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+ * The index would be used in DP table
+ *
+ */
+std::map<std::set<igraph_integer_t>, igraph_integer_t> subsetMap;
```
Using global variables like this is problematic. What if igraph is used from multiple threads? That could be fixed by adding `IGRAPH_THREAD_LOCAL`, but there are other problems too. What if an error occurs in the funtcion, or the computation is interrupted? In such cases memory should be released, and in C++ the best way to do it is with RAII. Ideally this should be refactored to be a local variable, possibly in the main Steiner tree function.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+			return it->second;
+		}
+	}
+	return -1;
```
Is the return value of `-1` meaningful, and is this line expected to ever be executed? If not, it should be changed to an assertion, e.g.:

`IGRAPH_FATAL("Must never reach here.");`

In more typical cases we use `IGRAPH_ASSERT` instead. It works the same wat as `assert()` but it is not disabled in production code.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+			return it->first;
+		}
+	}
+	return std::set<igraph_integer_t>();
```
Same comment as with the `-1` return from `fetchIndexofMapofSets()`. Is this expected to ever be executed?

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+	// Initially the value of m is the vertex that was removed from Steiner Terminals
+	igraph_integer_t m = q;
+
+	int len = C.size();
```
Let's stick to `igraph_integer_t` unless there is a good reason to use `int`. Keep in mind that `int` will be 32-bit on basically all supported platforms while `igraph_integer_t` will be 64-bit on 64-bit systems.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+
+	if (!IsConnected)
+	{
+		*res = 0.0;
+		res_tree = nullptr;
```
When an error occurs, do not set the value of output parameters.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+	if (igraph_vector_int_size(steiner_terminals) == no_of_vertices)
+	{
+
+		igraph_error_t result  = igraph_minimum_spanning_tree(graph, res_tree, weights);
```
Use `IGRAPH_CHECK` here instead of saving the error code.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+	 *	and returns distance matrix. The Dreyfus - Wagner algorithm needs complete graph
+	 *   hence this step is necessary.
+	 */
+	igraph_distances_johnson(graph, &distance, igraph_vss_all(), igraph_vss_all(), weights);
```
Make sure that all functions that return error codes are wrapped in `IGRAPH_CHECK()`, both on this line and elsewhere.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+	*res = distance2;
+
+	std::set<igraph_integer_t> newSet;
+	for (auto i = 0; i < igraph_vector_int_size(&steiner_terminals_copy); i++)
```
`auto` takes on the type of the initialized here, which is `0`, thus it has type `int`. All integers that are used to index into igraph vectors should be `igraph_integer_t`.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+ *	The algorithm uses Dynamic Programming approach and it assumes the graph is undirected.
+ *
+ * </para><para>
+ * Reference:
+ * S.E Dreyfus, R.A Wagner
+ * The steiner problem in graphs
+ * Networks Journal, 1971
+ * https://doi.org/10.1002/net.3230010302
```
```suggestion
 * Reference:
 * </para><para>
 * S.E Dreyfus, R.A Wagner
 * The Steiner problem in graphs
 * Networks 1, 195-207 (1971).
 * https://doi.org/10.1002/net.3230010302
```

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+	return IGRAPH_SUCCESS;
+}
+
+/* \function  igraph_steiner_dreyfus_wagner
```
Can you update the documentation to match the current function interface?

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,650 @@
+ * \sa \ref igraph_distances_johnson(), generateSubsets(), fetchIndexofMapofSets(), generate_steiner_tree_appx()
+ */
+
+igraph_error_t igraph_steiner_dreyfus_wagner(const igraph_t *graph, const igraph_vector_int_t *steiner_terminals,
```
Include `core/exceptions.h` and add `IGRAPH_HANDLE_EXCEPTIONS_BEGIN` and `IGRAPH_HANDLE_EXCEPTIONS_END` to the start and end of this function. This will make sure that C++ exceptions cannot escape into C code.

szhorvat reviewed (changes requested):
This is not a full review, but I added a few small technical comments that immediately stood out.

Nachiket18 commented:
> Before starting a review, I would like the code formatted in the same style as the rest of igraph. You can do this most easily with the `astyle` utility. An `.astylerc` file is included in the repository. If you're not sure how to do this, let me know and I will push that change.

Okay. Thanks. We were able to format using astyle utility.

szhorvat commented:
Could you please add the missing `IGRAPH_CHECK`s? Every function that returns an error code should have an `IGRAPH_CHECK`. That includes almost all igraph functions, with a few exceptions such as destructors or functions that return results directly such as `igraph_is_directed()`.

Nachiket18 commented:
> Could you please add the missing `IGRAPH_CHECK`s? Every function that returns an error code should have an `IGRAPH_CHECK`. That includes almost all igraph functions, with a few exceptions such as destructors or functions that return results directly such as `igraph_is_directed()`.

Thanks. @RonakGSahu made the changes and pushed.

RonakGSahu commented:
Hello @szhorvat @ntamas, just wondering if there is any update on the PR as it has been a week since we haven't received any communication on the review.

szhorvat reviewed (changes requested):
We were busy with the release of igraph 0.10, so there was not a lot of time to look at this PR yet. I will be away next week, and I'd like to ask you to make a few changes during that time.

I brought the PR up to date with the `master` branch. With the release of 0.10, development has moved from `develop` to `master`. I have also included this function in the main documentation. You will need to pull these changes before continuing.

 - Can you please go through all the old unresolved comments I left in the source code, check if there is anything left to address, and if not, mark them as resolved? It's difficult to keep track of what has been completed if this is not done. If you have any doubts about marking as resolved, then please ask.

 - Instead of rejecting disconnected graphs, please check if all Steiner terminals are in the same component. This will have to be done eventually anyway.

 - Implement interruption support. Include `core/interruption.h` and invoke `IGRAPH_ALLOW_INTERRUPTION()` in loops that may take a long time to complete. This function will check if the user requested interruption (when using igraph from R/Python/Mathematica) and if yes, it returns from the function. It should only be called from functions that return an `igraph_error_t`. Choosing where to do this check requires some though: it does have some performance cost, so it's not always a good idea to call from the innermost loops, but it _should_ be called often (ideally several times a second) so that interruption works in a responsive way.

    I plan to integrate this function into a high-level interface and test its functionality from there. It's important to have interruption support in place, otherwise interactive testing is impossible.

 - `std::map<std::set<igraph_integer_t>, igraph_integer_t>` and `std::set<igraph_integer_t>` are used frequently throughout the file. Can you create typedefs for them to make the code more readable?

 - Add a test for the single-vertex graph.

Nachiket18 commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,601 @@
+     */
+    IGRAPH_CHECK(igraph_distances_johnson(graph, &distance, igraph_vss_all(), igraph_vss_all(), weights));
+
+    /*
+        Setting distance from vertex to itself as 0.
+    */
+    for (igraph_integer_t i = 0; i < no_of_vertices; i++) {
+        if (MATRIX(distance, i, i) != 0) {
+            MATRIX(distance, i, i) = 0;
+        }
+    }
```
If there are self loops which is identified if the distance matrix has non-zero value of i'th row and i'th column then just mark it as zero

szhorvat replied:
If there are no negative weights (which you check for) then the diagonal of the shortest path matrix will contains only zeros. If you encountered a situation where this is not the case, that is a bug, so please report it.

Nachiket18 replied:
Okay. So no we didn't encounter any issues with the johnson's algorithm function so I have removed this piece of code in my latest commit.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,603 @@
+
+static igraph_integer_t fetchIndexofMapofSets(int_set subset, const dictionary& subsetMap) {
+
+    for (const auto & it : subsetMap) {
```
The `it` variable here is not an iterator, but the actual element stored in the map (a key-value pair). Can you please choose a different name than `it`?

szhorvat commented:
Thanks for the fixes! Next step is for me to integrate this function into igraph's Mathematica interface so that I can test the functionality and verify results. I'll take a little while before I can find the time to do this.

Nachiket18 commented:
@szhorvat - I'm facing an issue with creating a disconnected graph, when I create a graph and isolate a specific vertex in the test case, it is showing me 'Segmentation Fault' error. Could you help me with understanding how a disconnected graph is created in igraph? I'm using igraph_small function for creating graph of 7 vertices and while preparing the edges, I create edges for all vertices except for the 6th one.

szhorvat commented:
@Nachiket18 Can you paste the `igraph_small()` code here? Then I can let you know why it is crashing.

Here is an example that created a disconnected graph. Vertex 6 (i.e. the 7th vertex) has no edges.

```c
igraph_small(&g, 7, IGRAPH_UNDIRECTED,
    0,1, 1,2, 2,3, 3,4, 4,5,
    -1);
```

The most common reasons for crashes with `igraph_small()` are forgetting the `-1` at the end, or forgetting the comma before the `-1`.

Nachiket18 commented:
> @Nachiket18 Can you paste the `igraph_small()` code here? Then I can let you know why it is crashing.
> 
> Here is an example that creates a disconnected graph. Vertex 6 (i.e. the 7th vertex) has no edges. Furthermore, there are no connections between then 0-1-2 and 3-4-5 components.
> 
> ```c
> igraph_small(&g, 7, IGRAPH_UNDIRECTED,
>     0,1, 1,2, 3,4, 4,5,
>     -1);
> ```
> 
> The most common reasons for crashes with `igraph_small()` are forgetting the `-1` at the end, or forgetting the comma before the `-1`.

Thanks. We were able to figure out the issue which related to test file not using CHECK_ERROR() function.

Nachiket18 commented:
What would be next steps from our end? Should we create a new PR for approximate algorithm which would be based on the Dreyfus-Wagner algorithm?

szhorvat commented:
Sorry about taking so long on this. I'll try to finish the review on the weekend. There are a few more things to clean up here, but it should be straightforward.

szhorvat commented:
Just a heads up that I'll be pushing a few commits to this PR on the weekend (in case you have any local commits you haven't pushed yet).

szhorvat commented:
I'll be leaving some "FYI" comments as I go. There's no need to act on these at this moment.

----

The `IGRAPH_VECTOR_INT_INIT_FINALLY()` and similar macros are useful for initializing a vector and immediately putting it on the finally stack.

----

"Steiner" is a given name so it should be capitalized. "Tree" is not, so it shouldn't (except at the start of sentences).

----

Instead of

```
for (igraph_integer_t i=0; i < igraph_vector_size(vec); i++)
```

please always write

```
igraph_integer_t n = igraph_vector_size(vec);
for (igraph_integer_t i=0; i < n; i++)
```

in order to avoid calling `igraph_vector_size()` in a loop. Since this function cannot be inlined, it can have a real performance impact. It's easier to avoid this pattern consistently than to benchmark.

----

`igraph_vector_min()` must not be called on zero-length vectors. This is relevant when checking that all edge weights are valid. It must be checked whether the weight vector has a non-zero size.

----

In order to check if all terminals are in the same component, simply iterate through them, and check that they all have the same component ID. It is not necessary to have a nested loop.

----

See the above commit for reference.

szhorvat commented:
`IGRAPH_CHECK()` should only be used in functions that return an error code. This means that it must not be used in the `main()` function of tests.

szhorvat reviewed (changes requested):
Please check the last commit I made. It looks like initializing the result vectors to zero length instead of unit length causes the function to return a wrong result. Can you look into this and fix it?

Additionally:

 - Please re-use the result variables in the test
 - Print out both the vector and total weight in both tests
 - Add a couple more test cases on the same graph, if possible
 - Make sure that the text printed with the `printf`s is accurate

szhorvat commented:
It would be nice to also add a verification to the tests that checks that the result is indeed a tree.  This can be done by using `igraph_subgraph_edges()` to extract the result as a subgraph, and checking that it is a tree with `igraph_is_tree()`. All the checks, including printing the result, can be collected into a single function that is called for each test graph and its Steiner tree result.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,574 @@
+    }
+    igraph_integer_t indexD = fetchIndexofMapofSets(newSet, subsetMap);
+
+    igraph_vector_int_t vectorlist_all;
+    IGRAPH_VECTOR_INT_INIT_FINALLY(&vectorlist_all, 0);
```
This is a list of vertices, not a list of vectors, so it should have been named accordingly. However, it does not appear to be used at all. If not needed, please remove it.

Nachiket18 replied:
This is used to store the result which are vertices in the Steiner tree and is to be used as output(Output is length of Steiner tree and the tree itself). We are changing the name as you've suggested.

szhorvat replied:
The vertices of the tree are not at the moment returned by the top-level function. First, we need to decide whether it makes sense to return them. Currently, my opinion is that this is not necessary, as the vertices (as opposed to edges) would be rarely needed. If they are needed for some special application, they can be computed from the edges. Do you agree?

If the vertices are not returned or used anyway, then there is no need to record them, so this variable can be removed.

Nachiket18 replied:
Understood. I have removed vectorlist_all from as input parameter to the function calculating tree and it's references inside the function

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,574 @@
+ *  Populates the subsetMap data structure as well.
+ *
+ */
+static std::set<int_set> generateSubsets(igraph_vector_int_t steinerTerminals, igraph_integer_t n, igraph_integer_t graphsize, dictionary& subsetMap) {
```
Vectors should always be passed by reference, not by value as here.

szhorvat replied:
I fixed this.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,574 @@
+
+            IGRAPH_FINALLY_CLEAN(4);
+
+            int_set min_F;
+            min_F.insert(F1);
```
Here you create a new data structure, the modify it, but it immediately goes out of scope and gets destroyed.

Is there a mistake here?

Nachiket18 replied:
min_F is defined inside the while loop and it's used in if-else block. min_F is set to value D and D is the entry condition for the while loop which is why we have defined it.

szhorvat replied:
My point is that you defined two distinct variables named `min_F`. The one you are inserting into here is not the same as the one in the while loop.

Can you check this for mistakes?

Nachiket18 replied:
My apologies for the silly mistake. I have removed the code that re defines min_f inside the else block.

szhorvat commented:
My understanding was that the algorithm works for arbitrary real weights. Can you confirm this? In this case weights should be stored as reals, not integers (see the above commit). Could you please add some tests with non-integer weights? The current test uses only integral quantities.

szhorvat commented on src/paths/steiner.cpp:
```diff
@@ -0,0 +1,574 @@
+                        /*
+                         *  A set with Singleton value E removed from subset D
+                         */
+                        for (auto iter :  DMinusE) {
+                            if (iter == E) {
+                                iter = DMinusE.erase(iter);
+                                break;
+                            }
+                        }
```
The variable called `iter` here is not an iterator, but a value from `DMinusE`. Can you please change the name to a more meaningful.

Why do you assign the return value of `erase()` to `iter`?

RonakGSahu replied:
We changed the name. I think at that time if we didn't have a variable to store the return value from that function it would throw an error and quit the compilation but now it doesn't seem to happen so we have pushed what it was originally supposed to be (without the assignment).

szhorvat commented:
I'll assemble a TODO list below, to make it clear what we need to do to proceed.

 1. Do I understand it correctly that the algorithm works for non-integer weights? If so, can you please add some test cases with non-integer weights?
 2. Does the current implementation work when some weights are exactly zero? We need to figure this out and update the input checks accordingly.
 3. Can you please add some test cases which involve some non-trivial graph structure? Currently, the function is only tested on a complete graph.
 4. Can you please add some test cases which involve non-simple graphs? There should be self-loops and parallel edges with distinct weights. Be sure to verify that the function returns the lowest weight edge out of several parallel edges.
 5. It would be quite useful if the function would be able to interpret `weights == NULL` as all-unit weights. This feature would also make it easier to integrate this function into high-level igraph interface (Python, R, Mathematica). Can you please implement this? As a start, it's easiest if the top-level function checks whether `weights == NULL` and if yes, it generates an all-unit weight vector internally, and uses that later.

Nachiket18 commented:
> I'll assemble a TODO list below, to make it clear what we need to do to proceed (in addition to the above review comments).
> 
> 1. Do I understand it correctly that the algorithm works for non-integer weights? If so, can you please add some test cases with non-integer weights?
> 2. Does the current implementation work when some weights are exactly zero? We need to figure this out and update the input checks accordingly.
> 3. Can you please add some test cases which involve some non-trivial graph structure? Currently, the function is only tested on a complete graph.
> 4. Can you please add some test cases which involve non-simple graphs? There should be self-loops and parallel edges with distinct weights. Be sure to verify that the function returns the lowest weight edge out of several parallel edges.
> 5. It would be quite useful if the function would be able to interpret `weights == NULL` as all-unit weights. This feature would also make it easier to integrate this function into high-level igraph interface (Python, R, Mathematica). Can you please implement this? As a start, it's easiest if the top-level function checks whether `weights == NULL` and if yes, it generates an all-unit weight vector internally, and uses that later.

I started implementing the functionality to handle case number 5 with a simple piece of code below. Had to change the function signature by removing const for weights parameter as it's getting modified. 

if (!weights){
        //igraph_vector_int_resize(weights,no_of_edges);
        igraph_vector_init(weights,no_of_edges);
        igraph_vector_fill(weights,1);
}

Upon executing the following code in test case we are getting segmentation fault.

CHECK_ERROR(igraph_steiner_dreyfus_wagner(&g_k7, &terminals_k7, NULL, &value_unit, &tree_edges_unit), IGRAPH_EINVAL);

szhorvat commented:
It crashes because you are passing a `NULL` pointer to `igraph_vector_init()`. Of course, this cannot work. It leads to dereferencing a `NULL` pointer. Since it is not pointing to any usable memory area, no vector can be initialized there.

But it seems you misunderstood what I meant. The `weights` argument must never be modified by this function, and should be `const`. The goal is that when one passes `NULL` to the function, it should treat all weights as being one. How this is achieved is an implementation detail, and should be of no concern to the caller.  The simplest way to implement this behaviour is to internally allocate a vectors of ones, and use that _instead_ of the `weights` parameter.

Here's a possible pattern that has been used elsewhere in the library:

```c
igraph_error_t some_function(const igraph_vector_t *weights) {
    igraph_vector_t *pweights; // work with this throughout the function
    igraph_vector_t iweights; // this will be a vector of all-ones when weights == NULL

    if (! weights) {
        IGRAPH_VECTOR_FINALLY_INIT(&iweights, no_of_edges);
        igraph_vector_fill(&iweights, 1.0);
        pweights = &iweights;
    }

    // always use 'pweights' to access the weights

    if (! weights) {
        igraph_vector_destroy(&iweights);
        IGRAPH_FINALLY_CLEAN(1);
    }

    return IGRAPH_SUCCESS;
}
```

Let me know if you have more questions.

Nachiket18 commented:
> 

Thanks for the response. I have made the modification as you suggested for the the case where the weights are passed as NULL. The test case returns the value for Steiner Tree length as 3.

Nachiket18 commented:
2 . The algorithm assumes that all edge weights are positive. I double checked it in research paper.
3. Added two test cases with non-complete graphs and it's generating correct results.
5. Implemented the code for unit vector of weights and tested it with test case returning correct results.

szhorvat commented:
Thanks for your work on this! Just let us know when the PR is ready for the next round of reviews.

RonakGSahu commented:
Sorry for the delay, we were busy with schoolwork. We wrote the test case for real-valued weights but somewhere in the code (Data Structures), it is getting changed to weird values while `igraph_vector_t` by default is a real-valued vector and it should be the same. For example, with the same test. the value in the vector list is getting changed. Below is an example.

Weights that we initialized: 
```2.5, 2.5, 2.5, 0.5, 0.5, 2.5, 2.5, 2.5, 2.5, 0.5, 2.5, 2.5, 2.5, 2.5, 0.5, 0.5, 2.5, 0.5, 2.5, 0.5, 0.5```
Weights that the main function takes in
```
Trial 1: 0 0 1 1.59279e+09 0 0 0 0 0 0 0 0 0 0 0 0 0 -1 257 7 1.01505e+08 
Trial 2: 0 0 1 -1.22579e+09 0 0 0 0 0 0 0 0 0 0 0 0 0 -1 257 7 1.57788e+08 
Trial 3: 0 0 1 1.39985e+09 0 0 0 0 0 0 0 0 0 0 0 0 0 -1 257 7 1.42535e+08 
```
So I am currently thinking that there is a bug that is happening in the `vector_t` object because this weight list is being used everywhere in the code and for integer values it behave correctly as expected.

Nachiket18 commented:
` Can you please add some test cases which involve non-simple graphs? There should be self-loops and parallel edges with distinct weights. Be sure to verify that the function returns the lowest weight edge out of several parallel edges. ` - I'm working on this test cases and I wanted to know what function should be used to create a non simple graph. Is there a parameter for igraph_small function that can be used to allow it to accept non-simple graphs? @szhorvat

szhorvat commented:
`igraph_small()` does allow multi-edges and self-loops. For example, the following would create two parallel edges between `0` and 1`1`, three edges between `2` and `3` and a self-loop on `3`.

```
igraph_small(&graph, 0, IGRAPH_UNDIRECTED,
    0,1, 0,1, 1,2, 2,3, 2,3, 2,3, 3,3,
    -1);
```

What `igraph_simplify()` does is that it merges parallel edges and removes self-loops.

The question is whether the algorithm, as currently implemented, works for non-simple graphs. Self-loops are irrelevant for Steiner trees, so they should just be ignored. When there is more than one edge between two vertices, only the one with the smallest weight should be considered.

Does the current implementation do this correctly? Many algorithms generalize naturally to non-simple graphs. For example, the Dijkstra shortest implementation in igraph works fine for such graphs without needing to make special accommodations. But some don't.

If the current Steiner tree implementation does work for non-simple graphs, there should be tests for this case.  If it does not, the question is: can it be to work, without much effort? If not, such graphs should be rejected by the function.

RonakGSahu commented:
the question in general is GitHub still down? I resolved the problem when Nachiket raised it but I haven't been able to push since then.

szhorvat commented:
Is the issue resolved now? If you want to move to a new branch / new PR, that's fine.

Nachiket18 commented:
> `igraph_small()` does allow multi-edges and self-loops. For example, the following would create two parallel edges between `0` and 1`1`, three edges between `2` and `3` and a self-loop on `3`.
> 
> ```
> igraph_small(&graph, 0, IGRAPH_UNDIRECTED,
>     0,1, 0,1, 1,2, 2,3, 2,3, 2,3, 3,3,
>     -1);
> ```
> 
> What `igraph_simplify()` does is that it merges parallel edges and removes self-loops.
> 
> The question is whether the algorithm, as currently implemented, works for non-simple graphs. Self-loops are irrelevant for Steiner trees, so they should just be ignored. When there is more than one edge between two vertices, only the one with the smallest weight should be considered.
> 
> Does the current implementation do this correctly? Many algorithms generalize naturally to non-simple graphs. For example, the Dijkstra shortest implementation in igraph works fine for such graphs without needing to make special accommodations. But some don't.
> 
> If the current Steiner tree implementation does work for non-simple graphs, there should be tests for this case. If it does not, the question is: can it be to work, without much effort? If not, such graphs should be rejected by the function.

Thanks. I was making mistake in test case code that's why it was throwing error but I have rectified it now. The paper doesn't mention anything about parallel edges but I based on my understanding the algorithm should work correctly for parallel edges as well. Before starting the main algorithm, we pass it through Dijkstra's algorithm so that should be fine. I wrote a test case where I added a parallel edge which has higher length and it gave the same result as the one without parallel edge.

Nachiket18 commented:
> Is the issue resolved now? If you want to move to a new branch / new PR, that's fine.

It's working now. Thanks

szhorvat commented:
I mean pushing to github. Is it working fine now?

RonakGSahu commented:
> I mean pushing to github. Is it working fine now?

I am still not able to push to the repository. I don't know about Nachikets's Repository.

szhorvat commented:
I'm not sure what's wrong. I wonder if it's related to the fact that the work was done on the `develop` branch in @Nachiket18's fork, yet we also regularly push to `develop` in the main igraph repo, as it serves as the base for the next major version. I'm not familiar enough with git to give advice here. @ntamas, any ideas?  Otherwise, it would be best to open a new PR from a new branch (not named `develop`). The new branch should be forked from the latest version of `master`.

RonakGSahu commented:
I don't think the name should be a problem because, from my understanding, it is just changing where the fetch and merge's start point/end point difference while this error happened right after @Nachiket18 's last commit and the later `fetch upstream`. Also, after this GitHub had issues with Codespace so I am again guessing our repository got corrupted due to it

ntamas commented:
No ideas from my end either; maybe it makes sense to start from a clean slate by opening a new PR?

RonakGSahu commented:
I was able to push right now and I PR'd it to Nachiket. Looks like they finally fixed it.

Nachiket18 commented:
> I was able to push right now and I PR'd it to Nachiket. Looks like they finally fixed it.

We are ready fir next round of reviews. @szhorvat

szhorvat commented:
This PR has become a bit difficult to manage as it was opened directly from @Nachiket18's `develop` branch. It is good practice to base off a branch from the eventual target and work on that. Furthermore, in the meantime igraph development has shifted back from `develop` to `master` (as 0.10 was released).

This might be part of the reason why GitHub is not showing correctly the files that were changed: https://github.com/igraph/igraph/pull/1939/files

Would you be able to open an entirely new PR and move these changes there manually (to start with a clean slate again)? I suggest the following

 - Make sure that the `master` branch in your fork is in sync (and up to date with) the upstream igraph repo.
 - Create a new branch based off of `master`. You could name it e.g. `feat/steiner`: `git checkout -b feat/steiner`.
 - Manually copy over the files that you created or modified for this PR, then commit them. There's no reason to try to preserve the commit history, as the commits will eventually be squashed anyway.
 - Create a new PR from this new branch.

szhorvat commented:
Continued at https://github.com/igraph/igraph/pull/2288
